### PR TITLE
thunderFX: delegate autocast regions to thunder

### PR DESCRIPTION
**Pull request by kshitij12345**
This PR updates `splitter` for thunderFX to pass regions under autocast context manager to thunder (instead of inductor as it happens currently). 

In dynamo, these regions show up bounded within `torch.amp.autocast_mode._enter_autocast` and `torch.amp.autocast_mode._exit_autocast` for which we have corresponding symbols in thunder (see PR https://github.com/Lightning-AI/lightning-thunder/pull/1262 for more details on these symbols) .

Eg. Dynamo Graph with autocast enabled.
```python
def forward(self, L_x_: "f32[2, 2]"):
      l_x_ = L_x_
      
      # No stacktrace found for following nodes
      _enter_autocast = torch.amp.autocast_mode._enter_autocast('cuda:0', None, True, None)
      
       # File: /home/kkalambarkar/lightning-thunder/scratchpad/test.py:158 in func, code: return torch.matmul(x, x)
      matmul: "f16[2, 2]" = torch.matmul(l_x_, l_x_);  l_x_ = None
      
      # No stacktrace found for following nodes
      _exit_autocast = torch.amp.autocast_mode._exit_autocast(_enter_autocast);  _enter_autocast = _exit_autocast = None
      return (matmul,)
```

Also, it works if there is a split in the `torch.autocast` region. `fx.split_module` copies _enter_autocast and _exit_autocast for each split to correctly preserve the state.

Eg program
```python
def func(x):
    with torch.autocast(device):
        x = torch.matmul(x, x)
        # We will have a split because torch.sinc is autoregistered in thunder
        # (and it will be passed to inductor)
        x = torch.sinc(x)
        return torch.matmul(x, x)
```

Split Graph
```python
class GraphModule(torch.nn.Module):
    def forward(self, l_x_: "f32[2, 2]"):
        # No stacktrace found for following nodes
        thunder_1 = self.thunder_1(l_x_);  l_x_ = None
        inductor_2 = self.inductor_2(thunder_1);  thunder_1 = None
        thunder_3 = self.thunder_3(inductor_2);  inductor_2 = None
        return (thunder_3,)
        
    class thunder_1(torch.nn.Module):
        def forward(self, l_x_: "f32[2, 2]"):
            # No stacktrace found for following nodes
            _enter_autocast = torch.amp.autocast_mode._enter_autocast('cuda:0', None, True, None)
            
             # File: /home/kkalambarkar/lightning-thunder/scratchpad/test.py:158 in func, code: x = torch.matmul(x, x)
            x: "f16[2, 2]" = torch.matmul(l_x_, l_x_);  l_x_ = None
            
            # No stacktrace found for following nodes
            _exit_autocast = torch.amp.autocast_mode._exit_autocast(_enter_autocast);  _enter_autocast = _exit_autocast = None
            return x
      
    class inductor_2(torch.nn.Module):
        def forward(self, x: "f16[2, 2]"):
            # No stacktrace found for following nodes
            _enter_autocast = torch.amp.autocast_mode._enter_autocast('cuda:0', None, True, None)
            
             # File: /home/kkalambarkar/lightning-thunder/scratchpad/test.py:159 in func, code: x = torch.sinc(x)
            x_1: "f16[2, 2]" = torch.sinc(x);  x = None
            
            # No stacktrace found for following nodes
            _exit_autocast = torch.amp.autocast_mode._exit_autocast(_enter_autocast);  _enter_autocast = _exit_autocast = None
            return x_1
            
    class thunder_3(torch.nn.Module):
        def forward(self, x_1: "f16[2, 2]"):
            # No stacktrace found for following nodes
            _enter_autocast = torch.amp.autocast_mode._enter_autocast('cuda:0', None, True, None)
            
             # File: /home/kkalambarkar/lightning-thunder/scratchpad/test.py:160 in func, code: return torch.matmul(x, x)
            matmul_1: "f16[2, 2]" = torch.matmul(x_1, x_1);  x_1 = None
            
            # No stacktrace found for following nodes
            _exit_autocast = torch.amp.autocast_mode._exit_autocast(_enter_autocast);  _enter_autocast = _exit_autocast = None
            return matmul_1
```

### CODE REVIEW

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -5590,7 +5590,12 @@ def backward_autograd_function_apply(
     id="torch.amp.autocast_mode._enter_autocast",
     tags=(prims.OpTags.DONT_DCE, prims.OpTags.CTX_MANAGER_ENTER_EXIT_OP),
 )
-def autocast_enter(device_type, dtype=None, enabled=True):
+def autocast_enter(device_type, dtype=None, enabled=True, _unused_cache_enabled=True):
+    # We may receive device_type=cuda:0
+    # PyTorch applies autocast irrespective of device index.
+    # So, here we grab the device_type from the string.
+    device_type, unused_deviceno = devices._device_from_string_helper(device_type)
+    device_type = devices.devicetype_string(device_type)
```
Comment (kshitij12345) [inline thunder/torch/__init__.py:5598]: As per [docs](https://pytorch.org/docs/stable/amp.html#torch.autocast) `torch.autocast` takes `device_type`, but in practice it does take device + index eg. `cuda:0` and ignores the `index` part. Not sure if it was intended but for now, we mimic the behaviour.

Eg.

```python
import torch

device = 'cuda:1'
dtype = torch.float

x = torch.rand(2, 2, device=device, dtype=dtype, requires_grad=True)

def func(x):
    with torch.autocast("cuda:0"):  # NOTE: This is cuda:0
        return torch.matmul(x, x)

o = func(x)
print(o.dtype)  # torch.float16

```

File: thunder/dynamo/utils.py [hunk 2]
```diff
@@ -176,7 +176,21 @@ def try_execute_thunder_symbol(thunder_symbol: Symbol, node: torch.fx.Node) -> t
     """
     import thunder
     from thunder.core.trace import TraceCtx
+    from thunder.core.compile_data import compile_data_and_stats
+    from thunder.common import CompileData, CompileStats
 
+    # We require a corresponding `_enter_autocast` in trace to correctly pop the state,
+    # else we get error notifying that we are popping empty stack.
+    # As a work-around, we short circuit and return True.
```
Comment (kshitij12345) [inline thunder/dynamo/utils.py:184]: Is there a better way around this? @IvanYashchuk do you have any idea or suggestion?

Comment (IvanYashchuk) [reply thunder/dynamo/utils.py:184]: I suggest adding `if not get_compile_data().autocast_stack: return` before `.pop()` call in
https://github.com/Lightning-AI/lightning-thunder/blob/c5f8bf73f21f20be8e748bbf0801d62e69f4bd92/thunder/torch/__init__.py#L5630-L5631

**Review by IvanYashchuk — APPROVED** [review]

**Review by t-vi — APPROVED** [review]
Thank you @kshitij12345 @IvanYashchuk

**Comment by kshitij12345** [comment]
Failure is unrelated (torch.rms_norm consistency tests are failing), re-ran the pipeline but the problem persists. I have filed an issue https://github.com/Lightning-AI/lightning-thunder/issues/1395

**Pull request merged**